### DISCUSSION TITLE
Bug in to_sql - Vertica dialect

### DISCUSSION THREAD

**Discussion by leonidmakarovsky** [sqlalchemy/sqlalchemy#11135] [Usage Questions]
When I use a new version of SQLAlchemy 2.0.28 and vertica-python 1.3.8, I encounter the following bug:
TypeError: VerticaDialect.has_table() got an unexpected keyword argument 'info_cache'

My code:
df.to_sql(name = 'table',
                                    con=engine,
                                    schema = 'schema',
                                    if_exists = 'append', index = False,
                                    chunksize = None,
                                    method = None)

I investigated python3.12/site-packages/sqlalchemy/engine/reflection.py. Line 429 has the following:
        with self._operation_context() as conn:
            return self.dialect.has_table(
                conn, table_name, schema, info_cache=self.info_cache, **kw

However, the function definition indeed does NOT have info_cache as an argument. When I removed info_cache argument when calling the function, this work. But I can't be altering the official distribution of SQLAlchemy. Can you please fix the bug?

**Comment by zzzeek** [comment]
> Can you please fix the bug?

This is not our bug, this is a bug in the vertica dialect, has_table must accept ``**kw``:

https://github.com/sqlalchemy/sqlalchemy/blob/985193c407ffb891c8eed042fac6f9547a34d694/lib/sqlalchemy/engine/interfaces.py#L1802-L1809

**Comment by leonidmakarovsky** [comment]
Yes, has_table must accept **kw. However, the root cause of the bug is not **_**kw_**. The root cause of the bug is the extra parameter **_info_cache_** which does not belong there. When I removed **_info_cache_** parameter, it worked. I didn't touch **_**kw_**. Additionally, you just provided the definition of has_table in interfaces.py. It doesn't have **_info_cache_** parameter. Yet, you're calling the method passing **_info_cache_** parameter.

**Reply by CaselIT** [reply]
that method has `**kw`, so it's expected that any number of keyword could passed to that method.
The bug here is in the vertica dialect, not in sqlalchemy

**Reply by leonidmakarovsky** [reply]
The bug has already been reported here:
https://stackoverflow.com/questions/75281889/sqlalchemy-error-has-table-got-an-unexpected-keyword-argument-info-cache-en
But 2.0.1 does NOT fix it either. 
The vertica dialect has NOT changed. But the way you're calling this dialect has changed. This is the code from your older version. This works perfectly. So why did you introduce info_cache if dialects do NOT take it?

There is NO info_cache in 1.4.37:

    def has_table(self, table_name, schema=None):
        """Return True if the backend has a table of the given name.                                                                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                                                               
        :param table_name: name of the table to check                                                                                                                                                                                                                                                                                                                                          
        :param schema: schema name to query, if not the default schema.                                                                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                                                               
        .. versionadded:: 1.4 - the :meth:`.Inspector.has_table` method                                                                                                                                                                                                                                                                                                                        
           replaces the :meth:`_engine.Engine.has_table` method.                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                                                               
        """
        # TODO: info_cache?                                                                                                                                                                                                                                                                                                                                                                    
	with self._operation_context() as conn:
            return self.dialect.has_table(conn, table_name, schema)

**Reply by CaselIT** [reply]
`has_table` has always had `**kw` in the interface. This is 1.4.0: https://github.com/sqlalchemy/sqlalchemy/blob/e769ba4b00859ac8c95610ed149da4d940eac9d0/lib/sqlalchemy/engine/interfaces.py#L451

The fact that vertica dialect has not changed does not imply that there is no bug there

**Reply by leonidmakarovsky** [reply]
How can there be a bug there if it was perfectly working until you started calling their dialect with a new parameter? Did you coordinate with Vertica to introduce the new parameter before calling their dialect? Have you tested your code change with Vertica? As I said, I have the sqlalchemy 1.4.37 version LOCALLY and it does NOT have info_cache. I pasted the above code from the LOCAL sqlalchemy version 1.4.37 that I installed a while back. Anyway, I guess you guys are not going to fix the bug so I will have to dump SQLAlchemy and try to find the alternative ways to write to Vertica DB.

**Reply by CaselIT** [reply]
> How can there be a bug there if it was perfectly working until you started calling their dialect with a new parameter?

the dialect must implement the dialect interface, if a method has `**kw` in the interface then a dialect must support `**kw`. If it does not then it's a bug in the dialect.

It's on the vertica developers to test a particular version of sqlalchemy. They can specify that the dialect supports particular versions using requirements constraints. If the developers of vertica did not test on sqlalchemy 2 and do not support it they should have in the requirements `sqlalchemy<2`.

This is how every extension interface works.

**Comment by leonidmakarovsky** [comment]
Additionally, this is Vertica-sqlalchemy-dialect github. This has been there forever. There was no **kw and info_cache requirements. Can you also remove it please if it's not required for certain dialects?

https://github.com/vertica/vertica-sqlalchemy-dialect/blob/main/vertica_sqlalchemy_dialect/base.py

I raised the issue with Vertica as well, but clearly this is SQLAlchemy bug.

https://github.com/vertica/vertica-sqlalchemy-dialect/issues/15

**Reply by CaselIT** [reply]
> but clearly this is SQLAlchemy bug.

I disagree, the bug is in the dialect that does not follow the specified interface

**Reply by zzzeek** [reply]
the ``**kw`` has been there for many releases for years, however, there was a time when it wasn't there.  The vertica dialect copied the older version of the method that did not have ``**kw``.  Later, we added ``**kw``, Vertica did not update.  But then even later still (probably in 2.0), we started **using** the ``**kw`` , and that is why you see Vertica breaking now.

When you say "This has been there forever.", that's correct - **that's the bug** - nobody is maintaining sqlalchemy-vertica and updating it for newer interfaces.    All software must make updates when dependencies are upgraded.  that's the whole reason you can do things like [dependency pinning](https://pip.pypa.io/en/stable/reference/requirements-file-format/#requirements-file-format) in Python and basically every other language.  As dependencies change, software that uses them may become incompatible with those newer changes.

**Reply by leonidmakarovsky** [reply]
I raised it up with them. Thanks zzzeek!
